:toc:
:toclevels: 4
:sectnumlevels: 2
:sectnums:
:appversion: v1.0

= Arvados CWLtest

== Introduction

Common Workflow Language (CWL) is open and free workflow management system - mainly for computational biology. CWL is a main method to run workflows on Arvados platform. Arvados CWLtest library is dedicated to people, that develop pipelines in CWL and run them on Arvados. This library allows to create easy and reproducible tests for CWL scripts, which you can keep in your repository inside python script and run them using some python pytest testing package. Usage of this library requires some basic Python knowledge, but don't be affraid - is super easy. Lets get started.

Library contains some set of functions, that allows to run a process on Arvados and later get information about outputs and check if the files are as you expect. If you run the test from commandline following things happen:

1. Empty project is created on arvados 
2. `arvados-cwl-runner` creates a process that runs your CWL scripts inside this project
3. You can see the logs for the process
4. If workflow status is `completed` pytest pass. If workflow status is `failed` or `cancelled` your test will fail.
5. Optionally there are executed other tests on outputs from the run (you can see examples in later chapters)
6. After one week project including all processes will be removed from arvados. 

To know how to create tests via Python API check <<How to create tests from Python API>>.

Library contains some additional feature - arvados_synchronize_test_data. It is dedicated to teams that would like to perform tests on different arvados nodes (for example arkau, ardev, arind and other). For CWL pipelines you need to testing data. This functionality allows to sychronize testing data between arvados nodes. Instructions how to use this feature is described in <<Synchronize test data across Arvados instances>>.

== Installation

=== Set up conda environment

[source, bash]
----
wget https://repo.anaconda.com/miniconda/Miniconda3-py38_4.12.0-Linux-x86_64.sh
sh Miniconda3-py38_4.12.0-Linux-x86_64.sh
conda install --no-deps pip
pip install virtualenv
----

=== Create virtualenv in project directory

[source, bash]
----
virtualenv venv
----

=== Activate virtualenv

[source, bash]
----
source venv/bin/activate
----

=== You need to install Arvados Cwl Test via  'pip install', according to this instructions:

[source, bash]
----
pip install biocwl-test --extra-index-url https://__token__:<your_personal_token>@code.roche.com/api/v4/projects/34319/packages/pypi/simple
----

where `<your_personal_token>` is token with `read_api` scope that can be created on the GitLab in Preferences -> Access Tokens.


To install specific version use  `biocwl-test==<version>`. Example: `biocwl-test==0.2.dev16+g3858809`

In case of `ImportError: pycurl: libcurl link-time ssl backend (nss) is different from compile-time ssl backend (none/other)` error
fix it with:
[source, bash]
----
pip uninstall pycurl
pip install --compile  --no-cache-dir --install-option="--with-nss" pycurl==7.44
----

== Usage

Arvados Cwl Test is implemented as Python library and can be executed using pytest (https://docs.pytest.org/en/7.1.x/)

=== How to create tests from Python API

Example of usage::

[source, python]
----
from biocwltest.arvados_connection import utils


def test_single_step():

# run the test and define in variable test_run to use it later
    test_run = basic_arvados_test(
        biocwltest_testing_uuid,
        "Example test",
        "cwl_workflows/test_single_step/test_single_step.cwl",
        {
            "name": "example.txt"
            }
            )

# Chek if output collection is not empty - for example in cases when File[] or Directory[] is the output
    assert check_if_collection_output_not_empty(test_run)

# There is a repsresentation how output dictionary looks like
    assert create_ouputs_dict(test_run) == {
        'example.txt': {
            'size': 0,
            'basename': 'example.txt',
            'location': '240a2608b2d56bb36d2b3d00ae5fcf41+53/example.txt'
            }
            }
# Check if there is a specific file in outputs
    assert 'example.txt' in create_ouputs_dict(test_run)

# Be sure if output has some specific size you expect
    assert create_ouputs_dict(test_run)['example.txt]["size"] > 0

----
=== Variables used in multiple tests

Sometimes there are multiple testing scripts in single repository and there are some variables, that you would like to share between your python testing scripts, to not repeat them in every place. For this purpose Arvados CWL test contains implementation that allows to store them in json file named `variables.json`

For example::
[source, json]
----
{
  "testing_projects": {
    "ardev": "ardev-j7d0g-ucckjtjhhp7xq81",
    "arind": "arind-j7d0g-ky58se83cx2wh39",
    "arkau": "arkau-j7d0g-9cs24q86tesl6rm"
  },
  "resources": {
    "directories": {
      "two_1000000_inforR_fastq": {
        "class": "Directory",
        "path": "keep:271cbc530a4fe42173a72d53531ad849+225"
      }
    },
    "files": {
      "reference_genome": {
        "class": "File",
        "path": "keep:570c54e5cc295045cfe9f5b361d63e36+6185/Homo_sapiens_assembly38.fasta",
        "secondaryFiles": [
          {
            "class": "File",
            "path": "keep:570c54e5cc295045cfe9f5b361d63e36+6185/Homo_sapiens_assembly38.fasta.fai"
          }
        ]
      },
      "intervals": {
        "class": "File",
        "path": "keep:11a2a794048a689efb7ecb1e1e66d1e8+12334/wgs_calling_regions.hg38.bed"
      }
    }
  }
}
----

Use `FILES', 'DIRECTORIES` and `UUIDS` in python script importing them as::

[soource, python]
----
from biocwltest.arvados_connection.utils import FILES, DIRECTORIES, UUIDS

DIRECTORIES["two_1000000_inforR_fastq"]
UUIDS["akau"]
FILES["intervals"]

----

=== Execute Tests from Commanline

[source, bash]
----
pytest -s
----

To run single test define `-k keyword` to choose some subset of tests

[source, bash]
----
pytest -k my_lovely_test
----

To run all tests from specific file run:

[source, bash]
----
pytest test/test_main.py -s
----

To run more tests in parallel: 

[source, bash]
----
pytest -n 3
----

And more options you can find in pytest library documentation.

== Development of the library

. Fork or pull and create branch
. Write the code
- write unit tests for your functions
- build package (every commit builds package on Gitlab)
- merge request
